bip78/src/receiver/mod.rs

```diff
@@ -42,23 +53,91 @@ impl UncheckedProposal {
+    /// Receiver MUST check that the Original PSBT from the sender
+    /// can be broadcast, i.e. `testmempoolaccept` bitcoind rpc returns { "allowed": true,.. }
+    /// for `get_transaction_to_check_broadcast()` before calling this method.
+    ///
+    /// Check this if you generate bitcoin uri to receive PayJoin on sender request without manual human approval, like a payment processor.
+    /// Such so called "interactive" receivers are otherwise vulnerable to probing attacks.
+    /// If a sender can make requests at will, they can learn which bitcoin the receiver owns at no cost.
+    /// Broadcasting the Original PSBT after some time in the failure case makes incurs sender cost and prevents probing.
+    ///
+    /// Call this after checking downstream.
+    pub fn attest_tested_and_scheduled_broadcast(self) -> MaybeInputsOwned {
+        MaybeInputsOwned {
+            psbt: self.psbt,
+        }
+    }
+
+    /// Call this method if the only way to initiate a PayJoin with this receiver
+    /// requires manual intervention, as in most consumer wallets.
+    ///
+    /// So-called "interactive" receivers, like payment processors, that allow arbitrary requests are otherwise vulnerable to probing attacks.
+    /// Those receivers call `get_transaction_to_check_broadcast()` and `attest_tested_and_scheduled_broadcast()` after making those checks downstream.
+    pub fn attest_manual_receive_endpoint(self) -> MaybeInputsOwned {
+        MaybeInputsOwned {
+            psbt: self.psbt,
+        }
+    }
+}
+
+impl MaybeInputsOwned {
+    /// The receiver should not be able to sign for any of the Original PSBT's inputs.
+    /// Check that none of them are owned by the receiver downstream before proceeding.
+    pub fn input_script_pubkeys(&self) -> Vec<Result<&Script, RequestError>> {
+        todo!();
+    }
+
+    /// If the sender included inputs that the receiver could sign for in the original PSBT,
+    /// the receiver must either return error original-psbt-rejected or make sure they do not sign those inputs in the payjoin proposal.
+    ///
+    /// Call this after checking downstream.
+    pub fn attest_inputs_not_owned(self) -> MaybeScriptsSupported {
+        MaybeScriptsSupported {
             psbt: self.psbt,
         }
     }
+}
+
+impl MaybeScriptsSupported {
+    pub fn input_script_types(&self, network: bitcoin::Network) -> Vec<Option<AddressType>> {
+        todo!();
+    }
 
-    pub fn this_is_purely_interactive_wallet(self) -> UnlockedProposal {
-        UnlockedProposal {
+    /// If the sender's inputs are all from the same scriptPubKey type, the receiver must match the same type.
+    /// If the receiver can't match the type, they must return error unavailable.
+    ///
+    /// Call this after checking downstream.
+    ///
+    /// Note: mixed spends are not necessarily indicative of distinct wallet fingerprints but they can be.
+    /// This check is intended to prevent some types of wallet fingerprinting.
```

Comment:
I believe the check is actually an error in *the BIP itself* :) See https://github.com/bitcoin/bips/pull/1244

Reply:
I'm fine dropping the check for the time being, which I did. However, a 1.0 release we should match the spec (either they change or we change)

Reply:
I actually wanted to keep it according to spec the whole time. I just clarified what I meant.

Reply:
ok, I re-introduced this check

Reply:
Do you want it removed since we can do it internally, or do you want it as part of the checklist? based on your next comment, posted below, I think we should remove it

> We can actually do this check internally, no need to burden the consumer with it. The consumer only needs to get one address type which is the same for all inputs and check that one.
https://github.com/Kixunil/payjoin/pull/30#discussion_r952476032